lib/ruby_lsp/queue.rb

```diff
@@ -0,0 +1,189 @@
+
+    sig { params(request: T::Hash[Symbol, T.untyped]).void }
+    def push(request)
+      job = Job.new(request: request, cancelled: false)
+
+      # Remember a handle to the job, so that we can cancel it
+      @mutex.synchronize do
+        @jobs[request[:id]] = job
+      end
+
+      @job_queue << job
+    end
+
+    sig { params(id: T.any(String, Integer)).void }
+    def cancel(id)
+      @mutex.synchronize do
+        # Cancel the job if it's still in the queue
+        @jobs[id]&.cancel
+
+        # Cancel the job if we're in the middle of processing it
+        if @current_job&.request&.dig(:id) == id
+          @worker.raise(Cancelled)
+        end
+      end
+    end
+
+    sig { void }
+    def shutdown
+      # Close the queue so that we can no longer receive items
+      @job_queue.close
+      # Clear any remaining jobs so that the thread can terminate
+      @job_queue.clear
+      # Wait until the thread is finished
+      @worker.join
+    end
+
+    # Executes a request and returns a Queue::Result. No IO should happen in this method, because it can be cancelled in
+    # the middle with a raise
+    sig { params(request: T::Hash[Symbol, T.untyped]).returns(Queue::Result) }
+    def execute(request)
+      response = T.let(nil, T.untyped)
+      error = T.let(nil, T.nilable(StandardError))
+
+      request_time = Benchmark.realtime do
+        response = T.must(@handlers[request[:method]]).action.call(request)
+      rescue StandardError => e
+        error = e
+      end
+
+      Queue::Result.new(response: response, error: error, request_time: request_time)
+    end
+
+    # Finalize a Queue::Result. All IO operations should happen here to avoid any issues with cancelling requests
+    sig do
+      params(
+        result: Result,
+        request: T::Hash[Symbol, T.untyped]
+      ).void
+    end
+    def finalize_request(result, request)
+      error = result.error
+      if error
+        T.must(@handlers[request[:method]]).error_handler&.call(error, request)
+
+        @writer.write(
+          id: request[:id],
+          error: {
+            code: LanguageServer::Protocol::Constant::ErrorCodes::INTERNAL_ERROR,
+            message: result.error.inspect,
+            data: request.to_json,
+          },
+        )
+      elsif result.response != Handler::VOID
+        @writer.write(id: request[:id], result: result.response)
+      end
+
+      request_time = result.request_time
+      if request_time
+        @writer.write(method: "telemetry/event", params: telemetry_params(request, request_time, result.error))
+      end
+    end
+
+    private
+
+    sig { returns(Thread) }
+    def new_worker
+      Thread.new do
+        loop do
+          # Thread::Queue#pop is thread safe and will wait until an item is available
+          job = T.let(@job_queue.pop, T.nilable(Job))
+          # The only time when the job is nil is when the queue is closed and we can then terminate the thread
+          break if job.nil?
```

Comment:
Maybe using `while` can simplify this?

```rb
while job = T.let(@job_queue.pop, T.nilable(Job))
end
```

Reply:
`Thread::Queue#pop` automatically makes the thread sleep if there are no items in the queue and then wakes it up once there's a new item there.

It only returns `nil` when the queue has been closed, which allows us granular control on how to deterministically shutdown the worker.

Reply:
So we may do more work in the future than just call `break`?

Reply:
Yes that's what I meant 🙂